pkg/api/merge.go

```diff
@@ -0,0 +1,371 @@
+	}
+}
+
+func mergeProvisioningState(complete, partial ProvisioningState) ProvisioningState {
+	if partial == "" {
+		return complete
+	}
+	return partial
+}
+
+func mergeRouterProfiles(complete, partial []RouterProfile) []RouterProfile {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// preconditions: each RouterProfile in partial is named uniquely; each
+	// RouterProfile in complete is named uniquely
+
+	partialMap := make(map[string]RouterProfile, len(partial))
+	for _, p := range partial {
+		partialMap[p.Name] = p
+	}
+
+	rv := make([]RouterProfile, 0, len(complete))
+
+	for _, c := range complete {
+		if p, found := partialMap[c.Name]; found {
+			rv = append(rv, *mergeRouterProfile(&c, &p))
+			delete(partialMap, c.Name)
+		} else {
+			rv = append(rv, c)
+		}
+	}
+
+	for _, p := range partial {
+		if _, found := partialMap[p.Name]; found {
+			rv = append(rv, p)
+		}
+	}
+
+	return rv
+}
+
+func mergeRouterProfile(complete, partial *RouterProfile) *RouterProfile {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &RouterProfile{}
+	case partial == nil:
+		partial = &RouterProfile{}
+	}
+
+	return &RouterProfile{
+		Name:            mergeString(complete.Name, partial.Name),
+		PublicSubdomain: mergeString(complete.PublicSubdomain, partial.PublicSubdomain),
+		FQDN:            mergeString(complete.FQDN, partial.FQDN),
+	}
+}
+
+func mergeAgentPoolProfiles(complete, partial []AgentPoolProfile) []AgentPoolProfile {
+	if complete == nil && partial == nil {
+		return nil
+	}
+
+	// preconditions: each RouterProfile in partial is named uniquely; each
+	// RouterProfile in complete is named uniquely
+
+	partialMap := make(map[string]AgentPoolProfile, len(partial))
+	for _, p := range partial {
+		partialMap[p.Name] = p
+	}
+
+	rv := make([]AgentPoolProfile, 0, len(complete))
+
+	for _, c := range complete {
+		if p, found := partialMap[c.Name]; found {
+			rv = append(rv, *mergeAgentPoolProfile(&c, &p))
+			delete(partialMap, c.Name)
+		} else {
+			rv = append(rv, c)
+		}
+	}
+
+	for _, p := range partial {
+		if _, found := partialMap[p.Name]; found {
+			rv = append(rv, p)
+		}
+	}
+
+	return rv
+}
+
+func mergeAgentPoolProfile(complete, partial *AgentPoolProfile) *AgentPoolProfile {
+	switch {
+	case complete == nil && partial == nil:
+		return nil
+	case complete == nil:
+		complete = &AgentPoolProfile{}
+	case partial == nil:
+		partial = &AgentPoolProfile{}
+	}
+
+	return &AgentPoolProfile{
+		Name:         mergeString(complete.Name, partial.Name),
+		Count:        mergeIntPtr(complete.Count, partial.Count),
+		VMSize:       mergeVMSize(complete.VMSize, partial.VMSize),
+		VnetSubnetID: mergeString(complete.VnetSubnetID, partial.VnetSubnetID),
+		OSType:       mergeOSType(complete.OSType, partial.OSType),
+		Role:         mergeAgentPoolProfileRole(complete.Role, partial.Role),
+	}
+}
+
+func mergeIntPtr(complete, partial *int) *int {
```

Comment:
It's overwriting complete when partial is populated, seems like it could just be

```go
if partial != nil {
  return to.IntPtr(*partial)
} else if complete != nil {
  return to.IntPtr(*complete)
}
return nil